src/peft/tuners/lora/eva.py

```diff
@@ -0,0 +1,505 @@
+    Recursively apply a function to the module and its submodules.
+
+    If the module is a LoraLayer, the base layer is moved to the specified device. If the module has any LoraLayer
+    submodules, the function is applied recursively to each submodule. Otherwise, the module is moved to the specified
+    device.
+    """
+    if isinstance(module, LoraLayer):
+        module.base_layer.to(device)
+        return
+    if any(isinstance(submodule, LoraLayer) for submodule in module.modules()):
+        for child in module.children():
+            recursive_apply(child, device)
+    else:
+        module.to(device)
+
+
+def get_device_with_meta_params(model: torch.nn.Module) -> torch.device:
+    """
+    Get the device of the model's parameters. Useful if some parameters are on meta device.
+    """
+    devices = list({p.device for p in model.parameters() if str(p.device) != "meta"})
+    if len(devices) > 1:
+        raise ValueError("model has multiple devices")
+    return devices[0]
+
+
+def move_inputs_to_device(inputs, device: Union[str, torch.device]):
+    """
+    Move the inputs to the specified device. Adapted from hf.Trainer.
+    """
+    if isinstance(inputs, Mapping):
+        return type(inputs)({k: move_inputs_to_device(v, device) for k, v in inputs.items()})
+    elif isinstance(inputs, (tuple, list)):
+        return type(inputs)(move_inputs_to_device(v, device) for v in inputs)
+    elif isinstance(inputs, torch.Tensor):
+        return inputs.to(device)
+    else:
+        raise ValueError(f"unsupported input type: {type(inputs)}")
+
+
+def get_indices_fn_causal_lm(inputs: dict, config: LoraConfig):
+    """
+    if not all items in the input should be used for SVD, this function can be used to get the indices of the items
+    that should be used
+    """
+    mask = inputs.get("attention_mask", torch.ones_like(inputs["input_ids"])).bool()
+    if config.eva_config.use_label_mask and hasattr(inputs, "labels"):
+        mask = torch.logical_and(mask, inputs["labels"] != config.eva_config.label_mask_value)
+    return mask.nonzero()
+
+
+def forward_fn_dict(model, inputs):
+    return model(**inputs)
+
+
+@torch.no_grad()
+def get_eva_state_dict(
+    model: torch.nn.Module,
+    peft_config: LoraConfig,
+    dataloader: torch.utils.data.DataLoader,
+    forward_fn: Optional[callable] = forward_fn_dict,
+    get_indices_fn: Optional[callable] = get_indices_fn_causal_lm,
+    prepare_activations_fn: Union[callable, Dict[str, callable], None] = None,
+    show_progress_bar: bool = True,
+) -> dict:
+    """
+    Compute the SVD for each layer in the model.
+
+    This function computes the Singular Value Decomposition (SVD) for each layer in the model. It uses the incremental
+    PCA method to compute the SVD components. The function also checks for convergence of the computed components using
+    cosine similarity. The rank distribution for each layer is determined based on the explained variance ratio.
+
+    Parameters:
```

Comment:
You're using the numpydoc docstring convention here (which sklearn also uses), but we use the (google?) docstring convention in PEFT. Could you please adjust it here and throughout?

Reply:
done